src/services/MailService.ts

```diff
@@ -1,53 +1,94 @@
-import nodeMailer, { SentMessageInfo } from 'nodemailer';
-import BaseService from '@/services/BaseService';
+import { ConfigData } from '@/configuration';
+import logger from '@/helper/logger';
+import { PlatformMessage } from '@/interfaces/PlatformMessage';
+import QueueManager, { JobData } from './QueueManager';
+import { Job } from 'bull';
 import Mail from '@/interfaces/Mail';
-import PlatformMailTransporter from '@/interfaces/PlatformMailTransporter';
-import Utils from '@/utils';
+import { MailTransport } from './MailTransport';
 
-export default class MailService extends BaseService {
-	public constructor(name: string) {
-		super(name);
-	}
+const SERVICE_TYPE = 'mail';
 
-	protected async _send(transporter: PlatformMailTransporter, mail: Mail): Promise<SentMessageInfo> {
-		if (mail.attachments) {
-			const decodeFiles = (files: Array<{ content: any; filename: string }>) =>
-				files.map(({ content, filename }) => ({
-					filename,
-					content: Buffer.from(content, 'base64'),
-				}));
-			mail.attachments = decodeFiles(mail.attachments);
-		}
+export default class MailService {
+	private queueManager: QueueManager;
+	private platformIds: string[];
+	private transports: MailTransport[] = [];
 
-		const config = await Utils.getPlatformConfig(transporter.platformId);
-		if (config.mail.defaults.envelope) {
-			mail.envelope = {
-				from: config.mail.defaults.envelope.from || mail.from,
-				to: config.mail.defaults.envelope.to || mail.to,
-			};
-		}
+	/**
+	 *
+	 * @param queueManager
+	 * @param configuration
+	 */
+	constructor(queueManager: QueueManager, configuration: ConfigData[]) {
+		this.queueManager = queueManager;
+		this.platformIds = configuration.map((cfg) => {
+			queueManager.createQueue(SERVICE_TYPE, cfg.platformId, cfg.queue);
+
+			const mailOptions = Array.isArray(cfg.mail.options)
+				? (cfg.mail.options as ConfigData[])
+				: ([cfg.mail.options] as ConfigData[]);
+			mailOptions.forEach((options) => {
+				this.transports.push(new MailTransport(SERVICE_TYPE, cfg.platformId, { options, defaults: cfg.mail.defaults }));
+			});
 
-		return transporter.transporter.sendMail(mail);
+			return cfg.platformId;
+		});
 	}
 
-	protected _createTransporter(options: any, defaults: any): nodeMailer.Transporter {
-		// todo check default from becomes defined
-		return nodeMailer.createTransport(options, defaults);
+	/**
+	 *
+	 */
+	async startWorkers(): Promise<void> {
+		await Promise.all(
+			this.platformIds.map(async (platformId) => {
+				return this.queueManager.startWorker(SERVICE_TYPE, platformId, this.process.bind(this));
```

Comment:
add info log https://docs.hpi-schul-cloud.org/display/TSC/Error+Handling+and+Logging#ErrorHandlingandLogging-LoglevelsandLogsources 
what happens if a worker crashes?

Reply:
The logging happens within `QueueManager.startWorkers()`. I updated the log level from `debug` to `info`.

Reply:
Bull (the queue package) tries has the concept of **stalled jobs**: "When a job stalls, depending on the job settings the job can be retried by another idle worker or it can just move to the failed status." The workers are implemented asynchronously so that they do not block the Node event loop. As an alternative we can configure workers to run in seperate sandboxed processes. The number of concurrently running workers per queue is also configurable. 
https://optimalbits.github.io/bull/ (see "**Stalled Jobs**" and "**Sandboxed Processors**").